src/coreclr/src/jit/assertionprop.cpp

```diff
@@ -1959,12 +1958,48 @@ AssertionInfo Compiler::optAssertionGenJtrue(GenTree* tree)
     {
         std::swap(op1, op2);
     }
+
+    ValueNum op1VN = vnStore->VNConservativeNormalValue(op1->gtVNPair);
+    ValueNum op2VN = vnStore->VNConservativeNormalValue(op2->gtVNPair);
     // If op1 is lcl and op2 is const or lcl, create assertion.
     if ((op1->gtOper == GT_LCL_VAR) &&
         ((op2->OperKind() & GTK_CONST) || (op2->gtOper == GT_LCL_VAR))) // Fix for Dev10 851483
     {
         return optCreateJtrueAssertions(op1, op2, assertionKind);
     }
+    else if (vnStore->IsVNCheckedBound(op1VN) && vnStore->IsVNInt32Constant(op2VN))
```

Comment:
What kind of cases does this handle?
when op1 is a LclVar and op2 is a constant we use the if-then case above.
So I would guess that this would handle the case when op2 is a LclVar that is known to hold a constant.

Reply:
I looked it up in the source code:

checkedBound should really be called arrayLenth as it only contains ValueNumbers that represent an array.Lenth property.